File: batch_group.go
package microscope

// BatchTypeGroup groups related batch entries by watcher type.
type BatchTypeGroup struct {
	Type    EntryType `json:"type"`
	Label   string    `json:"label"`
	Entries []Entry   `json:"entries"`
}

var batchTypeOrder = []EntryType{
	TypeRequest,
	TypeQuery,
	TypeLog,
	TypeEvent,
	TypeNotification,
	TypeException,
	TypeCache,
	TypeRedis,
	TypeJob,
	TypeSchedule,
	TypeMail,
	TypeHTTPClient,
	TypeWebSocket,
	TypePerformance,
	TypeMetric,
	TypeCustom,
	TypeTopic,
}

func typeLabel(t EntryType) string {
	switch t {
	case TypeRequest:
		return "Requests"
	case TypeQuery:
		return "Queries"
	case TypeLog:
		return "Logs"
	case TypeEvent:
		return "Events"
	case TypeNotification:
		return "Notifications"
	case TypeException:
		return "Exceptions"
	case TypeCache:
		return "Cache"
	case TypeRedis:
		return "Redis"
	case TypeJob:
		return "Queue Jobs"
	case TypeSchedule:
		return "Scheduled Tasks"
	case TypeMail:
		return "Mail"
	case TypeHTTPClient:
		return "External Calls"
	case TypeWebSocket:
		return "WebSockets"
	case TypePerformance:
		return "Performance"
	case TypeMetric:
		return "Metrics"
	case TypeCustom:
		return "Custom Events"
	case TypeTopic:
		return "Redpanda Topics"
	default:
		return string(t)
	}
}

func groupBatchByType(batch []Entry) []BatchTypeGroup {
	byType := make(map[EntryType][]Entry, len(batchTypeOrder))
	for _, e := range batch {
		byType[e.Type] = append(byType[e.Type], e)
	}
	groups := make([]BatchTypeGroup, 0, len(batchTypeOrder))
	for _, t := range batchTypeOrder {
		entries := byType[t]
		if len(entries) == 0 {
			continue
		}
		groups = append(groups, BatchTypeGroup{
			Type:    t,
			Label:   typeLabel(t),
			Entries: entries,
		})
	}
	return groups
}

File: adaptor/python/microscope/batch_group.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from microscope.entry import Entry

BATCH_TYPE_ORDER = [
    "request",
    "query",
    "log",
    "event",
    "notification",
    "exception",
    "cache",
    "redis",
    "job",
    "schedule",
    "mail",
    "http-client",
    "websocket",
    "performance",
    "metric",
    "custom",
    "topic",
]

TYPE_LABELS = {
    "request": "Requests",
    "query": "Queries",
    "log": "Logs",
    "event": "Events",
    "notification": "Notifications",
    "exception": "Exceptions",
    "cache": "Cache",
    "redis": "Redis",
    "job": "Queue Jobs",
    "schedule": "Scheduled Tasks",
    "mail": "Mail",
    "http-client": "External Calls",
    "websocket": "WebSockets",
    "performance": "Performance",
    "metric": "Metrics",
    "custom": "Custom Events",
    "topic": "Redpanda Topics",
}


@dataclass
class BatchTypeGroup:
    type: str
    label: str
    entries: list[Entry]

    def to_dict(self) -> dict[str, Any]:
        return {
            "type": self.type,
            "label": self.label,
            "entries": [entry.to_dict() for entry in self.entries],
        }


def type_label(entry_type: str) -> str:
    return TYPE_LABELS.get(entry_type, entry_type)


def group_batch_by_type(batch: list[Entry]) -> list[BatchTypeGroup]:
    by_type: dict[str, list[Entry]] = {t: [] for t in BATCH_TYPE_ORDER}
    for entry in batch:
        by_type.setdefault(entry.type, []).append(entry)
    groups: list[BatchTypeGroup] = []
    for entry_type in BATCH_TYPE_ORDER:
        entries = by_type.get(entry_type, [])
        if not entries:
            continue
        groups.append(BatchTypeGroup(type=entry_type, label=type_label(entry_type), entries=entries))
    return groups
